feat: resolver rebased

itsyaasir opened the pull request:
# Description of change

This pull request includes several significant changes aimed at updating the resolver to the latest rebased IOTA.

Changes include:
- Updating all the dependencies
- Refactor for the tests


## Links to any relevant issues

<!-- Be sure to reference any related issues by adding `fixes #issue_number`. -->

## Type of change

<!-- Choose a type of change, and delete any options that are not relevant. -->

- Bug fix (a non-breaking change which fixes an issue)
- [x] Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation Fix

## How the change has been tested

<!-- Describe the tests that you ran to verify your changes. -->

<!-- Make sure to provide instructions for the maintainer as well as any relevant configurations. -->

## Change checklist

<!-- Tick the boxes that are relevant to your changes, and delete any items that are not. -->

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [ ] I have updated the CHANGELOG.md, if my changes are significant enough

UMR1352 commented on tests/did_resolution.rs:
:eye::eye: naughty :eye::eye:
You can't change a test just to make it work. When a DID is missing the right error to return is 404, if something goes wrong during resolution only then you return a 500.

itsyaasir replied:
I thought this was changed in the identity core package itself; but will investigate it more

UMR1352 reviewed (changes requested):
I'd probably just drop the use of `identity_iota::resolver::Resolver` in favor of a custom structure (`HashMap`??) to get the right client for the right network.
If `identity_iota_core::rebased::migration::identity::get_identity` is exposed by `identity_iota` I'd probably use that to resolve since it returns a `Result<Option<Identity>>` allowing you to easily handle 500 error as well as 404 (without having to look into the error message).

UMR1352 commented on Cargo.lock:
Do we have to track this file?

itsyaasir replied:
Not reallly! I might have added it accidentally

itsyaasir commented:
> I'd probably just drop the use of `identity_iota::resolver::Resolver` in favor of a custom structure (`HashMap`??) to get the right client for the right network. If `identity_iota_core::rebased::migration::identity::get_identity` is exposed by `identity_iota` I'd probably use that to resolve since it returns a `Result<Option<Identity>>` allowing you to easily handle 500 error as well as 404 (without having to look into the error message).

I have opted to use the `HashMap` structure; couple of things I changed was the did network tag to `unknwn` since the devnet was returning this `unknwn` but IIRC you are working on this so will change as soon as you are done.

itsyaasir committed:
feat: rework the resolver

UMR1352 commented:
@itsyaasir good job! Do you mind waiting a bit before merging this? RN we are reworking a bit DID resolution on Rebased and I feel these changes should account for that as well.

itsyaasir committed:
refactor: update network configuration and dependencies

- Updated the `.env.sample` to include new network configurations for testnet, devnet, and custom networks.
- Modified `Cargo.toml` to bump the `identity_iota` and `identity_storage` dependencies to version `v1.6.0-alpha.1`.
- Adjusted the `Dockerfile` to set the `NETWORK` environment variable to `devnet`.
- Enhanced the `README.md` to clarify network configuration options and provide examples for single and multiple network setups.
- Refactored `lib.rs` to implement a new `Network` enum for better handling of network configurations and client initialization.
- Updated tests to reflect changes in network handling and ensure compatibility with the new configurations.

UMR1352 commented on src/lib.rs:
```diff
@@ -69,93 +145,69 @@ pub struct ResolutionResponse {
-                    StatusCode::NOT_FOUND,
-                    "The requested DID document was not found".to_owned(),
-                )
-            }
-            _ => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
+        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
+        .ok_or_else(|| {
+            (
+                StatusCode::NOT_FOUND,
+                "The requested DID document was not found".to_owned(),
+            )
         })?;
```
Taking back what I said in a previous review comment. Unfortunately we can't just use `get_identity` as that only queries `OnChainIdentity`s and doesn't for instance resolve unmigrated aliases or migrated identities. We gotta go back to the previous implementation using `client.resolve_did` but making sure we return a 404 when the document is not found. 
To do that you gotta "parse" the error message you get and see if it matches the error you get when no DID document is found (although resolution didn't fail).

itsyaasir replied:
I have already redone it, please do take a look when you have the time.

itsyaasir committed:
feat use resolver

UMR1352 reviewed (approved):
Make sure the tests run succesfully

itsyaasir committed:
fix: handle request_faucet_funds result without error propagation

UMR1352 commented on tests/common/mod.rs:
```diff
@@ -96,7 +96,10 @@ pub async fn create_did(
 
     // Fund the account
-    request_faucet_funds(address).await?;
+    // WORKAROUND: Ignore the resullts of the faucet request as sometime the
+    // request will fail but the funds will be
+    // available.
+    let _ = request_faucet_funds(address).await;
```
Mmmm not sure if this is desirable... Can't we test on a localnet instead?

itsyaasir replied:
We can but we will have to publish the package which is not available here.

itsyaasir replied:
Anyone even what I did didn't work.

itsyaasir committed:
fix: handle faucet request result properly in create_did function

itsyaasir committed:
Update dependencies and improve IOTA sandbox setup with version selection support

itsyaasir committed:
Add --force-regenesis flag and update faucet endpoint URL

UMR1352 committed:
add mainnet support

UMR1352 committed:
use latest IOTA identity version

UMR1352 committed:
refactor error handling, new test

UMR1352 committed:
error handling, unknown network returns 400
